From 556c517783337bc88eae6632613f7f9e4ecb9f0d Mon Sep 17 00:00:00 2001
From: Elliot Chen <elliot.chen@nxp.com>
Date: Mon, 29 Nov 2021 17:15:07 +0800
Subject: [PATCH 50/91] MMFMWK-9010 alsasrc: add new feature of reading iec958
 stream

Upstream-Status: Pending
Signed-off-by: Elliot Chen <elliot.chen@nxp.com>
(cherry picked from commit fc97d964d8db2f00dc60c34c82ce007e6fc79d9f)
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 ext/alsa/gstalsa.c                      | 32 +++++++++++++++++++++----
 ext/alsa/gstalsasrc.c                   |  9 ++++++-
 gst-libs/gst/audio/audio-format.h       |  1 +
 gst-libs/gst/audio/gstaudioringbuffer.c |  3 ++-
 4 files changed, 39 insertions(+), 6 deletions(-)

diff --git a/ext/alsa/gstalsa.c b/ext/alsa/gstalsa.c
index 4d9b422..e005c4b 100644
--- a/ext/alsa/gstalsa.c
+++ b/ext/alsa/gstalsa.c
@@ -135,6 +135,8 @@ gst_alsa_get_pcm_format (GstAudioFormat fmt)
       return SND_PCM_FORMAT_FLOAT64_LE;
     case GST_AUDIO_FORMAT_F64BE:
       return SND_PCM_FORMAT_FLOAT64_BE;
+    case GST_AUDIO_FORMAT_IEC958_SUBFRAME_LE:
+      return SND_PCM_FORMAT_IEC958_SUBFRAME_LE;
     default:
       break;
   }
@@ -187,10 +189,16 @@ gst_alsa_detect_formats (GstObject * obj, snd_pcm_hw_params_t * hw_params,
     const GValue *format;
     GValue list = G_VALUE_INIT;
 
+    gboolean is_iec958_format = FALSE;
     s = gst_caps_get_structure (in_caps, i);
     if (!gst_structure_has_name (s, "audio/x-raw")) {
-      GST_DEBUG_OBJECT (obj, "skipping non-raw format");
-      continue;
+      if (gst_structure_has_name (s, "audio/x-iec958")) {
+        is_iec958_format = TRUE;
+        GST_DEBUG_OBJECT (obj, "audio/x-iec958");
+      } else {
+        GST_DEBUG_OBJECT (obj, "skipping non-raw format or iec958 format");
+        continue;
+      }
     }
 
     format = gst_structure_get_value (s, "format");
@@ -207,12 +215,28 @@ gst_alsa_detect_formats (GstObject * obj, snd_pcm_hw_params_t * hw_params,
         const GValue *val;
 
         val = gst_value_list_get_value (format, i);
-        if (format_supported (val, mask, endianness))
+        if (format_supported (val, mask, endianness)) {
           gst_value_list_append_value (&list, val);
+        } else {
+          if (is_iec958_format) {
+            if (snd_pcm_format_mask_test (mask,
+                    SND_PCM_FORMAT_IEC958_SUBFRAME_LE)) {
+              gst_value_list_append_value (&list, val);
+            }
+          }
+        }
       }
     } else if (G_VALUE_HOLDS_STRING (format)) {
-      if (format_supported (format, mask, endianness))
+      if (format_supported (format, mask, endianness)) {
         gst_value_list_append_value (&list, format);
+      } else {
+        if (is_iec958_format) {
+          if (snd_pcm_format_mask_test (mask,
+                  SND_PCM_FORMAT_IEC958_SUBFRAME_LE)) {
+            gst_value_list_append_value (&list, format);
+          }
+        }
+      }
     }
 
     if (gst_value_list_get_size (&list) > 1) {
diff --git a/ext/alsa/gstalsasrc.c b/ext/alsa/gstalsasrc.c
index 2e97f39..c5ddcfc 100644
--- a/ext/alsa/gstalsasrc.c
+++ b/ext/alsa/gstalsasrc.c
@@ -112,7 +112,11 @@ GST_STATIC_PAD_TEMPLATE ("src",
     GST_STATIC_CAPS ("audio/x-raw, "
         "format = (string) " GST_AUDIO_FORMATS_ALL ", "
         "layout = (string) interleaved, "
-        "rate = (int) [ 1, MAX ], " "channels = (int) [ 1, MAX ]")
+        "rate = (int) [ 1, MAX ], " "channels = (int) [ 1, MAX ]; "
+        "audio/x-iec958, "
+        "format = U32LE, "
+        "layout = (string) interleaved, "
+        "rate = (int) [ 1, MAX ], " "channels = (int) 2")
     );
 
 static void
@@ -742,6 +746,9 @@ alsasrc_parse_spec (GstAlsaSrc * alsa, GstAudioRingBufferSpec * spec)
     case GST_AUDIO_RING_BUFFER_FORMAT_TYPE_MU_LAW:
       alsa->format = SND_PCM_FORMAT_MU_LAW;
       break;
+    case GST_AUDIO_RING_BUFFER_FORMAT_TYPE_IEC958:
+      alsa->format = SND_PCM_FORMAT_IEC958_SUBFRAME_LE;
+      break;
     default:
       goto error;
 
diff --git a/gst-libs/gst/audio/audio-format.h b/gst-libs/gst/audio/audio-format.h
index c7aeef8..e847fdb 100644
--- a/gst-libs/gst/audio/audio-format.h
+++ b/gst-libs/gst/audio/audio-format.h
@@ -128,6 +128,7 @@ typedef enum {
   GST_AUDIO_FORMAT_F32BE,
   GST_AUDIO_FORMAT_F64LE,
   GST_AUDIO_FORMAT_F64BE,
+  GST_AUDIO_FORMAT_IEC958_SUBFRAME_LE,
   /* native endianness equivalents */
   GST_AUDIO_FORMAT_S16 = _GST_AUDIO_FORMAT_NE(S16),
   GST_AUDIO_FORMAT_U16 = _GST_AUDIO_FORMAT_NE(U16),
diff --git a/gst-libs/gst/audio/gstaudioringbuffer.c b/gst-libs/gst/audio/gstaudioringbuffer.c
index 1aee68f..af8f58e 100644
--- a/gst-libs/gst/audio/gstaudioringbuffer.c
+++ b/gst-libs/gst/audio/gstaudioringbuffer.c
@@ -249,7 +249,8 @@ gst_audio_ring_buffer_parse_caps (GstAudioRingBufferSpec * spec, GstCaps * caps)
       goto parse_error;
 
     spec->type = GST_AUDIO_RING_BUFFER_FORMAT_TYPE_IEC958;
-    info.bpf = 4;
+    gst_structure_get_int (structure, "channels", &info.channels);
+    info.bpf = 8;
   } else if (g_str_equal (mimetype, "audio/x-ac3")) {
     /* extract the needed information from the cap */
     if (!(gst_structure_get_int (structure, "rate", &info.rate)))
-- 
2.43.0

